### ISSUE TITLE
Customizing html template

### ISSUE THREAD

**Issue by pi0** [nuxt/nuxt#14195]
History: In Nuxt 2, we were supporting the ability to provide a [custom HTML template](https://github.com/nuxt/nuxt.js/blob/dev/packages/vue-app/template/views/app.template.html) used for the renderer. While it was simple, several issues:

- Only could be customized once by end-users (modules cannot hook to extend it and had to use other methods for injection)
- Can be outdated if we wanted to add a new placeholder for the built-in template
- Additional spaces or code format for customization can break hydration behavior
- With Nitro and moving to standalone server builds, we had to _convert_ it into a JS function
- Different behavior for SSR/SPA is not covered
- Runtime level customizations are not possible
- Opaque behavior without ability to inspect by Nuxt

This is the anatomy of the built-in template as a standard HTML structure: 

```html
<!DOCTYPE html>
<html {{ HTML_ATTRS }}>
  <head {{ HEAD_ATTRS }}>
    {{ HEAD }}
  </head>
  <body {{ BODY_ATTRS }}>
    {{ APP }}
  </body>
</html>
```

The main use-case of providing a custom template is extending placeholders to add a global static script or style or attributes. With Nuxt 3, I'm proposing to have a configuration interface in order to extend template:

```js
export default defineNuxtConfig({
  template: {
    htmlAttrs: [],
    headAttrs: [],
    headPrepend: [],
    headPrepend: [],
    bodyAttrs: [],
    bodyPrepend: [],
    bodyAppend: []
  }
})
```
(note: we shall make it consistent with [`RenderResult.meta`](https://github.com/nuxt/framework/blob/1b6289f60f70dee0f97f6516ea662e91297a0d3c/packages/nuxt/src/core/runtime/nitro/renderer.ts#L22) interface)

This interface is **forward-compatible**, **predictable**, **easily configurable** and **extandable by modules** and always leads to a valid HTML output.

More dynamic customization and runtime behavior can be provided by `nuxt:render` hooks (https://github.com/nuxt/nuxt.js/issues/14182). 

There was also an initial discussion to provide the ability completely override the template as an escape hatch. I believe unless it is proven there is a valid use case that cannot be supported by the Object interface, it is probably safer to not expose it. There might be cases we want to compile something that is not HTML (like a Native output?) but this probably deserves better refactor on renderer implementation to expose an API and allow full overriding renderer file, taking into account it is not rendering HTML.

**Comment by menuRivera** [comment]
For anyone looking for a way to include some **custom html attributes**, the solution that @pierres brings works like a charm!

```ts
// at server/plugins/html.render.ts
export default defineNitroPlugin((nitroApp) => {
  nitroApp.hooks.hook('render:html', (html, { event }) => {
    html.htmlAttrs.push("data-theme='light'")
  })
})
```
Result:
![image](https://user-images.githubusercontent.com/60537198/224124575-9583d2b2-9e23-425c-ad9a-12d7e69cc9a0.png)

**Comment by danielroe** [comment]
That hook will also be called when prerendering the `/200.html` page so you can still use it to customise your SPA loader page.

**Comment by rozbo** [comment]
> That hook will also be called when prerendering the `/200.html` page so you can still use it to customise your SPA loader page.

Ok! thanks a lot!

**Comment by smortexa** [comment]
Any update on this feature?

**Comment by manniL** [comment]
The feature already exists (see https://github.com/nuxt/framework/pull/6042) but lacks documentation

**Comment by danieldanielecki** [comment]
For those dealing just with GTM - there's a nowadays package https://github.com/zadigetvoltaire/nuxt-gtm, but it doesn't work for `AW-*` tags (https://github.com/zadigetvoltaire/nuxt-gtm/issues/15).

However, the solution with `server/plugins/html.render.ts` works completely fine.
It was new to me; it imports the plugin automatically, so I paste the docs for the future folks: https://nuxt.com/docs/guide/directory-structure/server#server-plugins.

**Comment by Norbz** [comment]
Hi there.
Im trying to migrate to nuxt 3 an old very specific usecase.

I am building small apps for a client that are then integrated into his website. I've build a tool that scrape an empty page from their website to use as a template. I used to inject the {{ APP }} into this empty page so I could test my dev in a simulated envirronement.

During export, I would dump some data in html in the app wrapper, and add some attributes for my client to we able to modify a few things (params, some labels etc) but that's another story almost.

This was originally made with Vue 2, then I moved it to nuxt to allow for easier development, but the lack of external html template makes it impossible to use. The solution with the render.html hook could be used (while really not being practical), but it doesn't work for dev environnement.

Is this feature completely out of the way ?
I know this is a far fetched use case

Thanks for your feedback
